### PR TITLE
NIFI-6462 ListHDFS should be triggerable

### PR DESCRIPTION
**Pull request by AxelSync**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Allows to trigger `ListHDFS` with an input `FlowFile`

<h3>Changes</h3>

-  `ListHDFS` accepts an input `FlowFile`
-  If input `FlowFile` is present, it's attributes are used during the evaluation of `DIRECTORY`
-  Input `FlowFile` attributes are transferred to output `FlowFile`s
-  input `FlowFile` is sent to `REL_ORIGINAL`

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 1]
```diff
@@ -74,7 +74,9 @@
         + "cases where files with the same timestamp are written immediately before and after a single execution of the processor. For each file that is "
         + "listed in HDFS, this processor creates a FlowFile that represents the HDFS file to be fetched in conjunction with FetchHDFS. This Processor is "
         +  "designed to run on Primary Node only in a cluster. If the primary node changes, the new Primary Node will pick up where the previous node left "
-        +  "off without duplicating all of the data. Unlike GetHDFS, this Processor does not delete any data from HDFS.")
+        +  "off without duplicating all of the data. Unlike GetHDFS, this Processor does not delete any data from HDFS."
+        +  "If the execution is triggered by an input flowfile, it's attribute will be transferred to the generated flowfiles, and it will be rooted to "
```
Comment (mgaido91) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]:
```suggestion
        +  "If the execution is triggered by an input FlowFile, its attributes will be transferred to the generated flowfiles, and it will be routed to "
```

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 2]
```diff
@@ -94,6 +96,7 @@
         + "problems. State is stored across the cluster so that this Processor can be run on Primary Node only and if a new Primary "
         + "Node is selected, the new node can pick up where the previous node left off, without duplicating the data.")
 @SeeAlso({GetHDFS.class, FetchHDFS.class, PutHDFS.class})
+
```
Comment (mgaido91) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]: please remove this

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 3]
```diff
@@ -172,6 +175,11 @@
         .description("All FlowFiles are transferred to this relationship")
         .build();
 
+    public static final Relationship REL_ORIGINAL = new Relationship.Builder()
+            .name("original")
```
Comment (mgaido91) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]: too many indent spaces, please see the others

**Comment by AxelSync** [comment]
cc @mgaido91

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 4]
```diff
@@ -292,31 +301,34 @@ public void onPropertyModified(final PropertyDescriptor descriptor, final String
...
-            }
+
         }
 
         return toList;
     }
 
     @Override
     public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+
```
Comment (mgaido91) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]: please remove this empty line

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 5]
```diff
@@ -392,15 +404,22 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
 
         for (final FileStatus status : listable) {
             final Map<String, String> attributes = createAttributes(status);
-            FlowFile flowFile = session.create();
-            flowFile = session.putAllAttributes(flowFile, attributes);
+            FlowFile flowFile;
+            if(isTriggered){
```
Comment (mgaido91) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]:
```suggestion
            if (isTriggered) {
```

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 6]
```diff
@@ -392,15 +404,22 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+
+
             session.transfer(flowFile, REL_SUCCESS);
 
             final long fileModTime = status.getModificationTime();
             if (fileModTime > latestTimestampEmitted) {
                 latestTimestampEmitted = fileModTime;
             }
         }
+        if(isTriggered) session.transfer(originalFlowFile, REL_ORIGINAL);
```
Comment (mgaido91) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]:
```suggestion
        if (isTriggered) {
            session.transfer(originalFlowFile, REL_ORIGINAL);
        }
```

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 7]
```diff
@@ -292,31 +301,34 @@ public void onPropertyModified(final PropertyDescriptor descriptor, final String
...
             long latestListingTimestamp = orderedEntries.lastKey();
 
-            // If the last listing time is equal to the newest entries previously seen,
-            // another iteration has occurred without new files and special handling is needed to avoid starvation
-            if (latestListingTimestamp == minTimestamp) {
-                // We are done if the latest listing timestamp is equal to the last processed time,
-                // meaning we handled those items originally passed over
-                if (latestListingTimestamp == latestTimestampEmitted) {
-                    return Collections.emptySet();
+                // If the last listing time is equal to the newest entries previously seen,
```
Comment (mgaido91) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]: it seems you only changes the indentation here...can you revert these change?

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 8]
```diff
@@ -392,15 +404,24 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
             final Map<String, String> attributes = createAttributes(status);
-            FlowFile flowFile = session.create();
-            flowFile = session.putAllAttributes(flowFile, attributes);
+            FlowFile flowFile;
+            if (isTriggered) {
+                flowFile = session.putAllAttributes(session.create(originalFlowFile), attributes);
+            } else {
+                flowFile = session.putAllAttributes(session.create(), attributes);
+            }
+
```
Comment (mgaido91) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]: nit: remove extra blank line

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java [hunk 9]
```diff
@@ -392,15 +404,24 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
 
         for (final FileStatus status : listable) {
             final Map<String, String> attributes = createAttributes(status);
-            FlowFile flowFile = session.create();
-            flowFile = session.putAllAttributes(flowFile, attributes);
+            FlowFile flowFile;
+            if (isTriggered) {
+                flowFile = session.putAllAttributes(session.create(originalFlowFile), attributes);
```
Comment (mgaido91) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java]: nit: here we can just have the creation of the `flowFile` and we can restore line https://github.com/apache/nifi/pull/3597/files#diff-4016ca1214c89d3709fb01745f784375L396 which is common in the 2 cases

**Comment by AxelSync** [comment]
Thanks for all the suggestions, I'm working on fixing them

**Comment by AxelSync** [comment]
@markap14 @toivoadams @rickysaltzer could you please review this PR?
Thanks

**Comment by mgaido91** [comment]
cc @bbende @jtstorck @markap14 may you please help reviewing this PR? Thanks.

**Comment by bbende** [comment]
ListHDFS is meant to maintain state for a specific directory and find newer files during each execution. This works because the processor knows it is tracking a single directory and can detect when the user changes the Directory property. By allowing incoming flow files it means the processor is now tracking a potentially infinite number of directories which are coming dynamically from incoming flow files.

This was discussed a few times on the mailing list and the general consensus was to leave ListHDFS and implement an alternate processor that supported incoming flow files, but did not maintain state, basically a one-time listing. I believe GetHDFSFileInfo already provides this functionality.

**Comment by mgaido91** [comment]
I see, I wasn't aware of that processor. Thanks @bbende! Sorry for the trouble. I think this should be closed then.

**Comment by AxelSync** [comment]
@bbende thanks for the advice, I wasn't aware of the dedicated processor.
This issue can be closed.
